### ISSUE TITLE
Question about - MeshBVH.deserialize performance for larger geometry

### ISSUE THREAD

**Issue by dbuck** [gkjohnson/three-mesh-bvh#213]
(Pardon, a contrived example stress test so far!)

While testing out some experiments with a workers example and 'excessive' detailed geometry, I see a spike on first render for boundsTree's generated in the worker and deserialized back on main thread. Three's projectObject -> intersectObjects -> computeBoundingSphere is turning out quite expensive for the first visible frame in this case. Generating the boundingBox/boundingSphere is built in already when creating via the constructor or compute paths, but not immediately clear if that should be handled in deserialize also.

![image](https://user-images.githubusercontent.com/983807/110200106-318d4480-7e2a-11eb-918d-f5d13c52c2a0.png)

I did a small hacky experiment on to see whether generating the boundingBox/boundingSphere on the tail end of deserialize would work, but I wasn't sure what the best route for that might be? (Still getting a feel for the layout of things!) bvh.traverse that builds the bounding box seems to be faster when deserializing a single chunk / single group geometry, but I'm unsure if there are other cases which would need to be considered?

![image](https://user-images.githubusercontent.com/983807/110200215-c98b2e00-7e2a-11eb-8952-f34c90c35aa2.png)

```
		// similar to the end of buildTree, we can fill in details of the
		// boundingBox/boundingSphere from the loaded bvh data if they did not already exist
		// otherwise it gets generated from the attributes on first render frame
		if ( setBounds && geometry.boundingBox == null ) {

			const rootBox = new Box3();
			geometry.boundingBox = new Box3();

			bvh.traverse( ( depth, isLeaf, boundingData ) => {

				// when should we *actually* stop traversal?
				if ( depth >= 1 ) return true;

				geometry.boundingBox.union( arrayToBox( boundingData, rootBox ) );

			} );

		}

		if ( setBounds && geometry.boundingSphere == null ) {

			geometry.boundingSphere = geometry.boundingBox.getBoundingSphere( new Sphere() );

		}
```

**Comment by gkjohnson** [comment]
> I see a spike on first render for boundsTree's generated in the worker and deserialized back on main thread.

The spike looks like it's from the garbage collector so the fact that it happened during bounding sphere creation is coincidence. GC will have to run at some point but hopefully garbage isn't created so much that a major GC has to occur. With the BVH being generated in the Worker I wouldn't expect extra memory created there to affect the main context GC so if I had to guess I'd think the extra memory is from creating the 10 large TorusKnotBufferGeometries. With the settings in your screenshot the buffer attributes alone take up ~31 megabytes of memory so with 10 you're up to 310 megabytes. Three.js also creates a lot of large temporary arrays when [creating TorusKnotBufferGeometry](https://github.com/mrdoob/three.js/blob/dev/src/geometries/TorusKnotGeometry.js) that will have to get GC'd which will be at least twice the size of the buffer arrays because JS uses 64 bit doubles for numbers so that's 620 megabytes that have to get cleaned up after generation. If you're interested in testing _just_ the BVH creation logic I would create just a single worst case buffer geometry and clone it rather than using the three.js constructor to avoid that.

Having said that there are still improvements to be made to memory allocation during BVH creation -- I just don't think it's related to this spike. If you're interested here's a brief overview:

Right now in master there are three ways a BVH can be generated. By default "lazyGeneration" is true and only the first couple nodes of the tree are created and child nodes are generated as needed however this leaves a lot of unnecessary objects in memory in order generate those subsequent children and the final bvh representation is a series of JS objects rather than a tightly packed buffer. With the right settings the full tree of objects can be generated immediately without packing to a buffer but of course the final memory footprint of the tree is much larger than the packed buffer representation.

And lastly the bvh can be tightly packed to a buffer in order to take up less memory or be transferred from a worker. However all the intermediate JS objects mentioned above must still be generated so a lot of temporary memory winds up being created. With [v0.4.0](https://github.com/gkjohnson/three-mesh-bvh/pull/205) creating the buffer version of the BVH will be the only option so it will probably be easier to get rid of the intermediate object representation when creating the BVH and put the data straight into the buffer. That's an issue for after the v0.4.0 PR gets merged, though.

**Comment by gkjohnson** [comment]
Thinking about it more I figured the bounding box was already being set when deserializing the BVH but I see it's not. When generating the BVH we set `geometry.boundingBox` [here](https://github.com/gkjohnson/three-mesh-bvh/blob/master/src/buildFunctions.js#L657-L670) from the root bounds since we've basically already computed the bounds but it's never set when we deserialize from a worker. Setting the `geometry.boundingBox` on deserialize like it is done in the snippet linked above would definitely be a good change. I see you've generated the bounding box by iterating over all nodes [here](https://github.com/dbuck/three-mesh-bvh/compare/lint-update...dbuck:worker-pool-test?expand=1#diff-88bb7c6b000958d7030722771524542da43649dad5b83fb5c8bbbbccf9d71b71R185-R196) but using the root bounds only should give exactly the same result and be much faster. Perhaps adding a `getBoundingBox` function to the BVH would be helpful here so it can be used in both cases.

It makes me wonder if there's a fast way to generate a reasonable bounding sphere from the BVH that could be provided. Getting the bounding sphere from the bounding box is fast but the resulting volume is much larger than it needs to be. I've got a couple ideas but I'm not sure how much faster they would wind up being compared to the `BufferGeometry.getBoundingSphere` function. Perhaps it could be generated in the worker and passed back to the main context, as well.

Thanks for taking a deeper look at the worker code! I haven't been using it a whole lot so it hasn't gotten much attention.

**Comment by dbuck** [comment]
Sweet, thanks for taking the time to explain some of the tradeoffs on lazy vs full + worker, it's super helpful, and the 0.4 buffer changes sound interesting.

I'll probably have a chance to look at a more real world dataset this week (3d scans, likely via 3d-tiles), I don't foresee that we'll be needing to generate for 20m faces all at once! Should get a feel for the real world lazy vs worker tradeoffs, and with some consideration on max depth, I think it's flexible enough to cover, and seems like it should beat out my current octree rig.

It was getting late, but yes, it seemed like it would be faster to leverage some of the bounds info that needed to already be generated for the boundsTree, rather than allowing three to parse through the entire buffer again to read it. It wasn't immediately clear how I should get that roots info out of the packed buffer though without more digging, hence the traverse as poc + question ! :)

**Comment by gkjohnson** [comment]
> Should get a feel for the real world lazy vs worker tradeoffs

You can also run the benchmark to see some of the differences. With lazy the performance implications for raycast aren't so great. The first few raycasts wind up being pretty slow because it just shifts when the BVH is generated. Between the bloated memory, maintenance cost of handling two different tree representations, and questionable performance benefits it just wasn't worth it. Imo worker is definitely the better choice for minimizing / removing the stall of generating the BVH.

> I'll probably have a chance to look at a more real world dataset this week (3d scans, likely via 3d-tiles)

I'll be curious to hear what you find and if you think anything else can be improved for your use case. In the 3d tiles renderer I've worked on the raycast function is modified to leverage the bounding hierarchy already stored in the 3d tiles format and also [includes support for a similar firstHitOnly mechansim](https://github.com/NASA-AMMOS/3DTilesRendererJS#group) like is provided in this library to stop traversal early, as well. I do wonder how much benefit you'll get using this library + 3d tiles because 3d tiles are already broken up pretty well and tile geometries usually don't contain that many triangles just by the nature of the format.

Can I ask what kind of work you're doing? It sounds like there's a lot of crossover in what we do 😁 and I always like hearing how people are using my work. I also created two issues for the get bounding box and get bounding sphere functions (#216, #217) if you're interested in helping with either of those -- the contributions are always appreciated!

**Comment by dbuck** [comment]
> Imo worker is definitely the better choice for minimizing / removing the stall of generating the BVH.

I think that will end up being true if the serialize / buildTree adjusts slightly so that it doesn't require importing a second copy of the full three package into the worker bundles. On a superficial glance, it looked like it should be change-able by passing around the index buffer and position buffer instead of the geometry in many (most?) places.

Bleh, though even just including buffer brings in color, which then brings in mathutils and hey, you get a bonus 50k of unrelated JS.. ack.

> Can I ask what kind of work you're doing?
Enabling [higher quality scan data](https://my.matterport.com/show/?m=ui3dfrQDqB2&sm=2&sr=-2.14,-.81,-2.29&sp=9.57,8.66,-5.8) than our choices from 2014 constrained us to ;)

**Comment by dbuck** [comment]
> I do wonder how much benefit you'll get using this library + 3d tiles because 3d tiles are already broken up pretty well

True, depends on the size of the geometry within the tiles, yet another tradeoff to balance for requests vs load times (!) The built in tree of the tileset seems like it will be helpful in a few scenarios. Sidetracks of sidetracks though ;) Thanks for the tips.

**Comment by gkjohnson** [comment]
> I think that will end up being true if the serialize / buildTree adjusts slightly so that it doesn't require importing a second copy of the full three package into the worker bundles.

My impression was that browsers were smart enough to only load a single copy of a shared dependency between workers and the main context. Is this not the case? Of course the build process would have to be configured to correctly share those dependencies. I haven't done in depth testing on this, though.

> Enabling higher quality scan data than our choices from 2014 constrained us to ;)

Oh awesome -- I'll be interested to see what you guys come up with!
